File: 587-two-sum-unique-pairs/two-pointers.cc
class Solution {
public:
    /**
     * @param nums: an array of integer
     * @param target: An integer
     * @return: An integer
     */
    int twoSum6(vector<int> &nums, int target) {
        // write your code here
        if (nums.empty()) {
            return 0;
        }
        sort(nums.begin(), nums.end());
        unordered_set<int> hashSet;
        int left = 0;
        int right = nums.size() - 1;
        int count = 0;
        while (left < right) {
            int sum = nums[left] + nums[right];
            if (sum == target) { 
                if (hashSet.find(nums[left]) == hashSet.end()) {
                    count++;
                    hashSet.insert(nums[left]);
                }
                left++;
                right--;
            }
            if (sum > target) {
                right--;
                
            }
            if (sum < target) {
                left++;
            }
        }
        return count;
    }
};

File: 384-longest-substring-without-repeating-characters/two-pointers.java
public class Solution {
    /**
     * @param s: a string
     * @return: an integer
     */
    public int lengthOfLongestSubstring(String s) {
        // write your code here
        Set set = new HashSet();
        if (s == null || s.length() == 0) {
            return 0;
        }
        char[] sc = s.toCharArray();
        int result = Integer.MIN_VALUE; 
        int i = 0;
        int j = 0;
        while (j < sc.length) {
            if (!set.contains(sc[j])) {
                set.add(sc[j]);
                j++;
                result = Math.max(result, j - i);
            } else {
                while (j < sc.length && i < sc.length && sc[i] != sc[j]) {
                    set.remove(sc[i]);
                    i++;
                }
                set.remove(sc[i]);
                i++;
            }
        }
        return result;
    }
}
